src/js/angular/guides/steps/complex/execute-sparql-query/plugin.js

```diff
@@ -37,57 +39,66 @@ PluginRegistry.add('guide.step', [
+                                .catch((error) => {
+                                    services.toastr.error(services.$translate.instant('guide.unexpected.error.message'));
+                                    throw error;
                                 });
-                        }),
+                        },
                         onNextValidate: () => {
-                            const editorQuery = GuideUtils.removeWhiteSpaces(window.editor.getValue());
-                            const stepQuery = GuideUtils.removeWhiteSpaces(query);
-                            if (editorQuery !== stepQuery) {
-                                if (editorQuery === 'select*where{?s?p?o.}limit100' || overwriteQuery) {
-                                    // The query is the default query OR we previously overwrote it => we can overwrite it
-                                    window.editor.setValue(query);
-                                } else {
-                                    GuideUtils.noNextErrorToast(toastr, $translate, $interpolate,
-                                        'guide.step_plugin.execute-sparql-query.query-not-same.error', options);
-                                    return false;
-                                }
-                            }
-                            overwriteQuery = true;
-                            return true;
+                            return YasguiComponentDirectiveUtil.getOntotextYasguiElementAsync(SPARQL_DIRECTIVE_SELECTOR)
+                                .then((yasgui) => yasgui.getQuery().then((query) => ({yasgui, queryFromEditor: query})))
+                                .then(({yasgui, queryFromEditor}) => {
+                                    const editorQuery = GuideUtils.removeWhiteSpaces(queryFromEditor);
+                                    const stepQuery = GuideUtils.removeWhiteSpaces(query);
+                                    if (editorQuery !== stepQuery) {
+                                        if (editorQuery === 'select*where{?s?p?o.}limit100' || overwriteQuery) {
+                                            // The query is the default query OR we previously overwrote it => we can overwrite it
+                                            yasgui.setQuery(query);
+                                        } else {
+                                            GuideUtils.noNextErrorToast(toastr, $translate, $interpolate,
+                                                'guide.step_plugin.execute-sparql-query.query-not-same.error', options);
+                                            return false;
+                                        }
+                                    }
+                                    overwriteQuery = true;
+                                    return true;
+                                });
                         },
                         initPreviousStep: () => new Promise((resolve, reject) => {
```

Comment:
This function is called by the previous step's functionality. The process is complex because each step doesn't have prior knowledge of other steps before or after it. When a user clicks on 'Previous,' the guide attempts to display the previous step, but there's a challenge in determining if the previous step can be displayed. For example, if the previous step requires displaying results from a Sparql query, the query must be executed before the step can be displayed. To address this, I implemented this function. It is called recursively for every previous step until it reaches a step that doesn't depend on previous ones. At this point, the step initializes itself and returns a promise. The next step goes through its own initialization process, which can involve loading a page, executing a query, or opening a menu. It then resolves its promise, and this chain continues until the algorithm reaches the step that needs to be displayed. At this moment, initialized correctly for the step to be displayed as in the right flow.